src/string/is_valid_email_address.rs

```diff
@@ -0,0 +1,303 @@
+    for i in 1..input.len() {
+        if input.chars().nth(i).unwrap() == '"' && input.chars().nth(i - 1).unwrap() != '\\' {
+            quote_count += 1;
+
+            if !new_local_part.starts_with('"') && quote_count != 1 {
+                new_local_part.push('"');
+            }
+        }
+
+        if quote_count % 2 == 0 {
+            new_local_part.push(input.chars().nth(i).unwrap());
+        }
+    }
+
+    new_local_part
+}
+
+fn has_bad_dots(part: &str) -> bool {
+    part.starts_with('.') || part.ends_with('.') || part.contains("..")
+}
+
+fn contains_legal_local_characters(local_part: &str) -> bool {
+    for item in local_part.chars() {
+        if !CHARACTERS.contains(item) && !SYMBOLS.contains(item) && item != '.' && item != '"' {
+            return false;
+        }
+    }
+    true
+}
+
+fn contains_legal_domain_characters(domain: &str) -> bool {
+    for item in domain.chars() {
+        if !CHARACTERS.contains(item) && item != '-' && item != '.' {
+            return false;
+        }
+    }
+    true
+}
+
+fn is_valid_quotes(part: &str) -> bool {
+    for i in 1..part.len() {
+        if part.chars().nth(i - 1).unwrap() == '"' && part.chars().nth(i).unwrap() == '"' {
+            let proceeding_quote = part.chars().nth(i + 1);
+            if proceeding_quote.is_some() && proceeding_quote.unwrap() != '.' {
+                return false;
+            }
+        }
+    }
+    true
+}
+
+fn is_ipv4(ip: &str) -> bool {
+    let parts = ip
+        .split('.')
+        .map(|x| x.to_string())
+        .collect::<Vec<String>>();
+    if parts.len() != 4 {
+        return false;
+    }
+
+    for value in parts {
+        let value = value.parse::<u32>();
+        if value.is_err() || value.unwrap() > 255 {
+            return false;
+        }
+    }
+
+    true
+}
+
+fn is_ipv6(ip: &str) -> bool {
+    let parts = ip
+        .split(':')
+        .map(|x| x.to_string())
+        .collect::<Vec<String>>();
+
+    // we check for the length 9 here as the IPv6 counts as one of the parts
+    if parts.len() != 9 || parts.first().unwrap() != "IPv6" {
+        return false;
+    }
+
+    for value in &parts[1..parts.len()] {
+        if value.is_empty() || value.len() > 4 {
+            return false;
+        }
+
+        for chr in value.chars() {
+            if !HEX_CHARACTERS.contains(chr.to_ascii_lowercase()) {
+                return false;
+            }
+        }
+    }
+
+    true
+}
+
+pub fn is_valid_ip(domain: &str) -> bool {
+    is_ipv4(domain) || is_ipv6(domain)
+}
+
+fn is_valid_local_part(local_part: &str) -> bool {
+    if local_part.len() > MAX_LOCAL_PART_LENGTH as usize {
+        return false;
+    }
+
+    if !is_valid_quotes(local_part) {
+        return false;
+    }
+
+    if !contains_legal_local_characters(local_part) {
+        return false;
+    }
+
+    if has_bad_dots(local_part) {
+        return false;
+    }
+
+    true
+}
```

Comment:
What do you think about having a similar approach as in `is_valid_domain`? I.e. if the `local_part` starts and ends `"`. then a function `is_valid_local_quoted_part` and `is_valid_local_unquoted_part`?

Reply:
I don't think this would work as not all email addresses are made up of only a single string, for example `a."b".c` is valid although it does not begin and end with quotes. In addition due to the fact that the contents of all strings are removed at the start of the local part checking, I believe that it would make the code less readable.

Reply:
`a."b".c` is _unquoted_. It contains a `"` character, therefore it is not valid (cf. [wiki](https://en.wikipedia.org/wiki/Email_address#Local-part)).